### ISSUE TITLE
Overlowing Redis connections

### ISSUE THREAD

**Issue by kurtostfeld** [mediocregopher/radix.v3#37]
Using the radix.v3 client, my Go app will sporadically overflow a Redis db running on localhost. My Go app is using a single radix.v3 Redis pool of size 500. In a redis-cli client I can see `connected_clients:10000` which is the max limit. I suspect all the redis connections are erroring out, the Go client is attempting to open new connections faster than Redis can handle. Redis thinks it's maxed out at 10k connections.

Restarting the Go app temporarily resolves the problem, but after a few hours, or maybe a day, the issue occurs again.

FYI, the Redis db is running on localhost. It's running the latest 4.0.11. Nothing else is using Redis except a manual redis-cli client.

Any ideas on how to resolve this?

**Comment by mediocregopher** [comment]
Hmmm that's very strange. Can you give me some sample code which shows how you're using radix? How are you instantiating your Pool?

**Comment by kurtostfeld** [comment]
I create the Redis pool:

```go
	redisPool, err := radix.NewPool("tcp", "localhost:6379", 500)
	if err != nil {
		log.Fatalf("Redis connection error: %v\n", err)
	}
	log.Printf("Redis connection redisPool created successfully.\n")
	defer redisPool.Close()
```

I use the Redis pool in exactly one location:

```go
        // queryList will have 1 or 2 keys in it.
	redisErr := redisPool.Do(radix.Cmd(&redisResults, "MGET", queryList...))
``

It's very simple overall.

**Comment by mediocregopher** [comment]
I attempted to reproduce this locally and wasn't able to. I used [this gist](https://gist.github.com/mediocregopher/8d40d4e5db3ebf42aa52420a2ec583a8) to do it. With a `parallel` value at or lower than 500 I didn't see the `connected_clients` increase passed 500 hardly at all. If I raise it higher than the pool size `connected_clients will climb up into the thousands, but I never saw higher than 8000.

What I would recommend is to review your code and ensure that, if you're running the `EXISTS` call inside a go-routine which is being created more than once, that the number of routines created is bounded (i.e. not being created infinitely), and that there's not expected to be more than 500 go-routines using the pool at any moment.

If you need more than 500 routines using the pool at any moment I recommend raising the pool size. Alternatively you can use `radix.PoolOnFullBuffer`, like this:

```
radix.NewPool("tcp", "localhost:6379", 500, radix.PoolOnFullBuffer(1000, 1*time.Second))
```

This will allow you to increase the of the pool up to 1500 total connections during high load, then resize back down to the normal 500 during normal load automatically.

**Comment by kurtostfeld** [comment]
This is one of two officially recommended Redis Go libraries (https://redis.io/clients#go). I tried switching to the other one (https://github.com/gomodule/redigo) with Vitess connection pools (https://godoc.org/vitess.io/vitess/go/pools), and my problem went away, my app is completely stable.

My app is super simple: ~50 million ipv4 keys with blank string values in Redis. A Go app that opens a Redis connection pool and does exactly one simple Redis EXIST query on every incoming IP addresses about 15-20k/sec.

When using Radix my open Redis connections stayed at 500, when I used Redigo, my open connections stays at about 7, even though I set maxCap to 500. Obviously, it looks like Radix opens all 500 connections up front, while Redigo opens as needed, which isn't a problem by itself.

When using the radix library, Redis would report 20k+ open connections when it went into the bad state, which I presume is the Radix connection pooling library going into a loop of abandoning old connections and opening new connections. I'd guess there is some glitch/problem with the Radix library connection pooling.

I'm probably just going to keep using Redigo for my current tasks, I really don't care which as long as it's stable + simple.

Huge thank you for responding to my issue and putting effort into attempting to reproduce my issue.
